Cleanup log4tango

t-b opened the pull request:
Close #728. 
Close #713.
Close #712.

t-b committed:
cppapi/server/logstream.cpp: Remove commented out code

t-b committed:
log4tango: Remove LOG4TANGO_LOGGERS_USE_LOGSTREAM functionality

This define is never enabled. So we also don't need it for tango.

By removing it we have less code to maintain.

t-b committed:
log4tango: Remove other build systems than CMake

t-b committed:
log4tango: Remove library and package code

We don't ship log4tango as separate library anymore.

t-b committed:
log4tango/include/config.h.in: Remove user facing configuration header

t-b committed:
log4tango: Remove LOG4TANGO_HAVE_NAMESPACES

Nested namespaces work nowadays out of the box with recent C++
compilers. We are also *not* using the result anywhere.

t-b committed:
log4tango/ChangeLog: Remove it

The last edit was from 2002...

t-b committed:
log4tango: Remove LOG4TANGO_HAVE_SSTREAM

We always have std::ostringstream with recent C++ compilers. In addition
defining classes in std namespace as fallback is undefined behaviour.

t-b committed:
log4tango/doc: Remove it

t-b committed:
log4tango: Remove the Windows specific configuration header

With CMake we can now use the generic one. It was also never installed
with CMake.

t-b committed:
config/config.h.in: Remove unused #cmakedefine parameters

t-b committed:
log4tangp: Remove DLLMain function for windows

We are not shipping log4tango as separate library anymore.

t-b committed:
log4tango: Reimplement threading utilities with standard C++11 features

We don't need to have platform specific solutions with pthread and
Windows threads. Instead we can just use the default
std::mutex, std::recursive_mutex and std::lock_guard classes.

t-b committed:
log4tango: Remove NDC code

It is not used in tango and would require us to implement thread local
storage.

So let's just remove it.

t-b committed:
log4tango: Remove empty PortabilityImpl files

t-b committed:
log4tango/Portability.hh: Remove warning disabling

This is bad style to do unconditionally and is also not tailored for our
current compilers.

mliszcz commented on log4tango/include/log4tango/threading/StandardThreads.hh:
```diff
@@ -1,37 +1,39 @@
-    
+using Mutex = std::mutex;
+using RecursiveMutex = std::recursive_mutex;
+using ScopedLock = std::lock_guard<std::mutex>;
```
I suggest to remove these definitions and use 'std' classes directly. Or we may adapt a rule that we always use these aliases instead of 'std' classes. This is to avoid possible confusion due to mixing different names.

* Mutex is referenced only once, from: AppenderAttachable.hh
* RecursiveMutex is never used
* ScopedLock is used in a few places in AppenderAttachable.cpp and Logger.cpp.

t-b replied:
Good point. Done.

mliszcz commented on log4tango/config/config.cmake:
```diff
@@ -27,32 +27,10 @@ endmacro()
-check_library_exists(pthread pthread_create "" HAVE_PTHREAD_CREATE)
-check_include_file("pthread.h;pthread_np.h" HAVE_PTHREAD_NP_H)
-
 if(CMAKE_THREAD_LIBS_INIT)
     set(LOG4TANGO_HAVE_THREADING "/**/")
 endif()
```
https://cmake.org/cmake/help/latest/module/FindThreads.html
> The thread library to use. This may be empty if the thread functions are provided by the system libraries and no special flags are needed to use them.

This check is probably bad for portability. Maybe we should check for `Threads_FOUND` instead?

Also, is it possible to use Tango without threads? After a quick check of omniORB source code it seems that they require threads (posix or ms or vxworks) and do not have any fallback implementation (compilation errors out with "Implementation header file incomplete"). And since log4tango is not standalone library, I suggest to do the following:
* require Threads_FOUND to be set here
* remove LOG4TANGO_HAVE_THREADING, remove DummyThread.hh/.cpp
* move log4tango/threading/StandardThreads.hh to log4tango/StandardThreads.hh if you decide to keep that file (I'd also rename to e.g. 'ThreadingUtils.hh)
* remove entire log4tango/threading directory

t-b replied:
Fair point. I've done that. And also reworked the `int64_t` check.

t-b replied:
@mliszcz I've created https://github.com/tango-controls/cppTango/issues/793 I think we should tackle that one in a separate PR. I don't see an issue with backwards compatibility.

mliszcz reviewed (changes requested):
LGTM! Very good initiative and I think we can use this opportunity to remove even more stuff.

t-b committed:
log4tango: Make threading support a requirement

Since log4tango is not used as a standalone library anymore, and one of
the tango dependecies, omniORB, always requires threading support, we
don't need to handle the case of no threading support here.

So let's require threading support.

Instead of having using declarations of the old types, we now also use
just the new std classes.

t-b committed:
log4tango/src/AppenderAttachable.cpp: Prefer std::make_pair

t-b committed:
log4tango/src/XmlLayout.cpp: Replace sprintf with std::to_string

This is less error prone and safer [1].

[1]: https://cwe.mitre.org/data/definitions/676.html

t-b committed:
log4tango/config: Remove unused configuration defines

We don't need to check if a certain header is present if the resulting
variable is never used.

And stdint.h, aka cstdint, is always available since C++11.

t-b committed:
log4tango/config/config.cmake: Make the int64_t check more generic

int64_t is optional in C++11. But if it is present and it's size is 8
bytes, we can use it, no matter what the platform or the OS bitness is.

Close #765.

mliszcz reviewed (approved):
LGTM! FYI in #765 I added a comment related to int64_t check because you are now resolving this one as well.
